src/core/lib/iomgr/exec_ctx.h

```diff
@@ -74,56 +75,87 @@ struct grpc_exec_ctx {
-    GRPC_CLOSURE_LIST_INIT, NULL, NULL, flags, gpr_cpu_current_cpu(),    \
-        finish_check_arg, finish_check, false, 0                         \
-  }
-
-/* initialize an execution context at the top level of an API call into grpc
-   (this is safe to use elsewhere, though possibly not as efficient) */
-#define GRPC_EXEC_CTX_INIT \
-  GRPC_EXEC_CTX_INITIALIZER(GRPC_EXEC_CTX_FLAG_IS_FINISHED, NULL, NULL)
-
 extern grpc_closure_scheduler* grpc_schedule_on_exec_ctx;
 
-bool grpc_exec_ctx_has_work(grpc_exec_ctx* exec_ctx);
+bool grpc_exec_ctx_has_work();
 
 /** Flush any work that has been enqueued onto this grpc_exec_ctx.
  *  Caller must guarantee that no interfering locks are held.
  *  Returns true if work was performed, false otherwise. */
-bool grpc_exec_ctx_flush(grpc_exec_ctx* exec_ctx);
+bool grpc_exec_ctx_flush();
 /** Finish any pending work for a grpc_exec_ctx. Must be called before
  *  the instance is destroyed, or work may be lost. */
-void grpc_exec_ctx_finish(grpc_exec_ctx* exec_ctx);
+void grpc_exec_ctx_finish();
 /** Returns true if we'd like to leave this execution context as soon as
     possible: useful for deciding whether to do something more or not depending
     on outside context */
-bool grpc_exec_ctx_ready_to_finish(grpc_exec_ctx* exec_ctx);
+bool grpc_exec_ctx_ready_to_finish();
 /** A finish check that is never ready to finish */
-bool grpc_never_ready_to_finish(grpc_exec_ctx* exec_ctx, void* arg_ignored);
+bool grpc_never_ready_to_finish(void* arg_ignored);
 /** A finish check that is always ready to finish */
-bool grpc_always_ready_to_finish(grpc_exec_ctx* exec_ctx, void* arg_ignored);
+bool grpc_always_ready_to_finish(void* arg_ignored);
 
 void grpc_exec_ctx_global_init(void);
 
 void grpc_exec_ctx_global_init(void);
 void grpc_exec_ctx_global_shutdown(void);
 
-grpc_millis grpc_exec_ctx_now(grpc_exec_ctx* exec_ctx);
-void grpc_exec_ctx_invalidate_now(grpc_exec_ctx* exec_ctx);
+grpc_millis grpc_exec_ctx_now();
+void grpc_exec_ctx_invalidate_now();
 gpr_timespec grpc_millis_to_timespec(grpc_millis millis, gpr_clock_type clock);
 grpc_millis grpc_timespec_to_millis_round_down(gpr_timespec timespec);
 grpc_millis grpc_timespec_to_millis_round_up(gpr_timespec timespec);
 
+inline grpc_exec_ctx make_exec_ctx(grpc_exec_ctx r) {
+  grpc_exec_ctx_flush();
+  return r;
+}
+
+class ExecCtx {
+ public:
+  ExecCtx();
+  ExecCtx(uintptr_t fl, bool (*finish_check)(void* arg),
+          void* finish_check_arg);
+  ~ExecCtx();
+
+  grpc_closure_list closure_list;
+  /** currently active combiner: updated only via combiner.c */
+  grpc_combiner* active_combiner;
+  /** last active combiner in the active combiner list */
+  grpc_combiner* last_combiner;
+  uintptr_t flags;
+  unsigned starting_cpu;
+  void* check_ready_to_finish_arg;
+  bool (*check_ready_to_finish)(void* arg);
+
+  bool now_is_valid;
+  grpc_millis now;
+
+ private:
+  ExecCtx* last_exec_ctx;
+};
+
+extern thread_local ExecCtx* exec_ctx;
```

Comment:
Remove this in preference to an explicit Get method on ExecCtx

Reply:
We also need to check that thread_local is available everywhere we are.... IIRC shipping OSX or iPhone compilers don't provide support, meaning we may need to fall back to gpr_ library support

Reply:
Need to decide between a non inline Get and an exposed exec_ctx